File: code/jvm/repository/src/test/java/pt/isel/repositories/mem/UserRepositoryInMem.java
package pt.isel.repositories.mem;

import pt.isel.domain.security.PasswordValidationInfo;
import pt.isel.domain.users.User;
import pt.isel.repositories.users.UserRepository;

import java.util.ArrayList;
import java.util.List;

public class UserRepositoryInMem implements UserRepository {
    private final List<User> users = new ArrayList<>();
    private long nextId = 1;

    @Override
    public User create(String username, PasswordValidationInfo passwordValidationInfo) {
        User user = new User(nextId++, username, passwordValidationInfo);
        users.add(user);
        return user;
    }

    @Override
    public User findById(Long id) {
        return users.stream().filter(u -> u.id().equals(id)).findFirst().orElse(null);
    }

    @Override
    public User findByUsername(String username) {
        return users.stream().filter(u -> u.username().equals(username)).findFirst().orElse(null);
    }

    @Override
    public boolean hasUsers() {
        return !users.isEmpty();
    }

    @Override
    public List<User> findAll() {
        return new ArrayList<>(users);
    }

    @Override
    public void save(User entity) {
        users.removeIf(u -> u.id().equals(entity.id()));
        users.add(entity);
    }

    @Override
    public void deleteById(Long id) {
        users.removeIf(u -> u.id().equals(id));
    }

    @Override
    public void clear() {
        users.clear();
        nextId = 1;
    }
}

File: code/jvm/repository/src/test/kotlin/pt/isel/repositories/mem/UserRepositoryInMem.kt
package pt.isel.repositories.mem

import jakarta.inject.Named
import pt.isel.domain.User
import pt.isel.domain.auth.PasswordValidationInfo
import pt.isel.repositories.UserRepository

/**
 * Naif in memory repository non thread-safe and basic sequential id. Useful for unit tests purpose.
 */
@Named
class UserRepositoryInMem : UserRepository {
    private val users = mutableListOf<User>()

    override fun create(
        username: String,
        passwordValidationInfo: PasswordValidationInfo,
    ): User = User(users.size.toLong() + 1, username, passwordValidationInfo).also { users.add(it) }

    override fun findById(id: Long): User? = users.firstOrNull { it.id == id }

    override fun findByUsername(username: String): User? = users.firstOrNull { it.username == username }

    override fun hasUsers(): Boolean = users.isNotEmpty()

    override fun findAll(): List<User> = users.toList()

    override fun save(entity: User) {
        users.removeIf { it.id == entity.id }.apply { users.add(entity) }
    }

    override fun deleteById(id: Long) {
        users.removeIf { it.id == id }
    }

    override fun clear() {
        users.clear()
    }
}
